### DISCUSSION TITLE
About the inability to clear meta information of png format pictures

### DISCUSSION THREAD

**Discussion by Williamsu1116** [libvips/libvips#4224] [Q&A]
vips version is vips-8.15.2，I use the following command, vips resize /pdd/tmp/exiftool_adobe_after_effects.png out-1.jpeg[strip] 1.0
However, the creator field in the meta information still exists in the new image generated.

**Comment by jcupitt** [comment]
Hello @Williamsu1116,

It should work. I see:

```
$ vips copy k2.jpg x.jpg[strip]
$ vipsheader -a x.jpg
x.jpg: 1450x2048 uchar, 3 bands, srgb, jpegload
width: 1450
height: 2048
bands: 3
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 2.83465
yres: 2.83465
filename: x.jpg
vips-loader: jpegload
jpeg-multiscan: 0
jpeg-chroma-subsample: 4:2:0
```

There's no extra metadata.

**Reply by Williamsu1116** [reply]
But if I use the command to eliminate pictures in png format, there will be meta information that cannot be deleted. Currently, I need to operate the png pictures to eliminate the meta information.

**Reply by jcupitt** [reply]
It doesn't matter what source you use -- the metadata is removed by the saver.

Can you upload a sample image where this fails? What platform are you using?

**Reply by Williamsu1116** [reply]
Linux environment for docker containers

**Reply by Williamsu1116** [reply]
![exiftool_adobe_after_effects](https://github.com/user-attachments/assets/35de8a72-b1a4-48e9-b945-20d9f8a56c27)
This is the picture that failed to delete meta information

**Reply by kleisauke** [reply]
@jcupitt It seems the `vips_isprefix()`  arguments were reversed, likely due to confusion with `g_str_has_prefix()`. Proposed patch:
```diff
--- a/libvips/foreign/foreign.c
+++ b/libvips/foreign/foreign.c
@@ -1722,8 +1722,8 @@ vips_foreign_save_remove_metadata(VipsImage *image,
 	VipsForeignKeep keep = *((VipsForeignKeep *) user_data);
 
 	// we are only interested in metadata
-	if (!vips_isprefix(field, "png-comment-") &&
-		!vips_isprefix(field, "magickprofile-") &&
+	if (!vips_isprefix("png-comment-", field) &&
+		!vips_isprefix("magickprofile-", field) &&
 		strcmp(field, VIPS_META_IMAGEDESCRIPTION) != 0 &&
 		!g_str_has_suffix(field, "-data"))
 		return NULL;
```

(I can also submit a PR targeting the `8.16` branch if needed)

**Reply by jcupitt** [reply]
Oh heh that's bad. Sure, please make a PR on 8.16.

**Reply by kleisauke** [reply]
I just submitted PR #4225 for this.

**Comment by jcupitt** [comment]
... though I don't think that affects this case, it works fine for me:

```
$ vips resize exiftool_adobe_after_effects.png x.jpg[strip] 1.0
$ vipsheader -a x.jpg
x.jpg: 316x50 uchar, 3 bands, srgb, jpegload
width: 316
height: 50
bands: 3
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 2.83465
yres: 2.83465
filename: x.jpg
vips-loader: jpegload
jpeg-multiscan: 0
jpeg-chroma-subsample: 4:2:0
```

Perhaps you have some error in your filenames @Williamsu1116 ?

**Reply by kleisauke** [reply]
Try saving to `x.png[strip]`, I think PNG comments are removed if you save it as JPEG.

**Reply by jcupitt** [reply]
Ooops, I tried that, then checked the header of the wrong file. You're right, this needs a PR.

**Comment by jcupitt** [comment]
OK, fixed with #4225, this will be in 8.16, due in a few days. 

Thanks for the report, @Williamsu1116.

**Reply by Williamsu1116** [reply]
Doesn't this project also support meta information deletion of bmp format files?  After I deleted the file meta information, I found that the meta information of the image increased instead.

**Reply by jcupitt** [reply]
It should work, yes.

If there's a bug, we won't be able to fix it until we can reproduce it, and we can't reproduce it without a sample file and exact instructions. You need to give some more information.

**Reply by Williamsu1116** [reply]
Have you solved this bug and released the new features in version 8.16.0?

**Reply by Williamsu1116** [reply]
I am referring to the bug of clearing the meta information error of png format pictures.

**Reply by jcupitt** [reply]
Yes, 8.16.0 has this fix.

**Reply by Williamsu1116** [reply]
But still cannot delete the creator field information in the image's meta information

**Reply by Williamsu1116** [reply]
![exiftool_adobe_after_effects](https://github.com/user-attachments/assets/31c793f7-a217-4ef6-83b8-8bebdb30d660)
vips copy exiftool_adobe_after_effects.png 1105.png[strip]
  Here are my test images and instructions The meta information is still retained in the newly generated image 1105

**Reply by kleisauke** [reply]
> But still cannot delete the creator field information in the image's meta information

Are you sure you tested this with libvips 8.16?  I can't reproduce this:

<details>
  <summary>Details</summary>

```console
$ vipsheader -a exiftool_adobe_after_effects.png
exiftool_adobe_after_effects.png: 316x50 uchar, 4 bands, srgb, pngload
width: 316
height: 50
bands: 4
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 1
yres: 1
filename: exiftool_adobe_after_effects.png
vips-loader: pngload
png-comment-0-Creator: Adobe After Effects
xmp-data: 416 bytes of binary data
bits-per-sample: 8
$ vips copy exiftool_adobe_after_effects.png 1105.png[strip]
$ vipsheader -a 1105.png
1105.png: 316x50 uchar, 4 bands, srgb, pngload
width: 316
height: 50
bands: 4
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 1
yres: 1
filename: 1105.png
vips-loader: pngload
bits-per-sample: 8
```
</details>
